genie-web/src/main/java/com/netflix/genie/web/rpc/grpc/services/impl/v4/GRpcJobTerminationServiceImpl.java

```diff
@@ -0,0 +1,120 @@
+ */
+@GrpcService(
+    value = JobTerminationServiceGrpc.class,
+    interceptors = {
+        SimpleLoggingInterceptor.class,
+    }
+)
+@Slf4j
+public class GRpcJobTerminationServiceImpl
+    extends JobTerminationServiceGrpc.JobTerminationServiceImplBase
+    implements JobKillServiceV4 {
+
+    private final Map<String, StreamObserver<JobKillListenerResponse>> parkedJobKillResponseObservers =
+        Maps.newConcurrentMap();
+    private final JobSearchService jobSearchService;
+
+    /**
+     * Constructor.
+     *
+     * @param jobSearchService Job search service
+     */
+    public GRpcJobTerminationServiceImpl(final JobSearchService jobSearchService) {
+        this.jobSearchService = jobSearchService;
+    }
+
+    /**
+     * Register to be notified when a kill request for the job is received.
+     *
+     * @param request          Request to register for getting notified when server gets a job kill request.
+     * @param responseObserver The response observer
+     */
+    @Override
+    public void registerForKillNotification(
+        final JobKillListenerRequest request,
+        final StreamObserver<JobKillListenerResponse> responseObserver
+    ) {
+        parkedJobKillResponseObservers.put(request.getJobId(), responseObserver);
+    }
+
+    @Override
+    public void killJob(final @NotBlank(message = "No job id entered. Unable to kill job.") String jobId,
+                        final @NotBlank(message = "No reason provided.") String reason) throws GenieException {
+
+        final StreamObserver<JobKillListenerResponse> responseObserver =
+            cleanUpAndGetParkedJobKillResponseObserver(jobId);
+
+        if (responseObserver != null && !jobSearchService.getJobStatus(jobId).isFinished()) {
+            responseObserver.onNext(
+                JobKillListenerResponse.newBuilder().build()
+            );
+            responseObserver.onCompleted();
+
+            log.info("Agent notified for killing the job ");
+        } else if (responseObserver == null) {
+            log.error("Job not killed. No response observer found for killing the job with id - {} ", jobId);
+            throw new GenieServerException(
+                "Job not killed. No response observer found for killing the job with id - {} " + jobId);
+        } else {
+            log.info("v4 job {} was already finished when the kill request came", jobId);
+        }
+    }
+
+    private StreamObserver<JobKillListenerResponse> cleanUpAndGetParkedJobKillResponseObserver(final String jobId) {
+
+        final StreamObserver<JobKillListenerResponse> responseObserver = parkedJobKillResponseObservers.remove(jobId);
+
+        //Note that this is possible in race conditions like job was finished so a clean up happened.
+        //When a disconnect happens then a clean up will happen because heart beat service will also send
+        //a notification. At this point there would be no response observer in the parkedJobKillResponseObservers map
```

Comment:
If we don clean up only on agent's disconnection then this will also not happen

Reply:
I think this whole method can go, no need for this comment or these INFO messages